release/releaser.go

```diff
@@ -1,279 +1,51 @@
 package release
 
 import (
-	"fmt"
 	"time"
 
 	"github.com/go-kit/kit/log"
-	"github.com/go-kit/kit/metrics"
 
-	"github.com/weaveworks/flux"
-	"github.com/weaveworks/flux/cluster"
-	fluxmetrics "github.com/weaveworks/flux/metrics"
-	"github.com/weaveworks/flux/policy"
-	"github.com/weaveworks/flux/registry"
 	"github.com/weaveworks/flux/update"
 )
 
-type ServiceUpdate struct {
-	ServiceID     flux.ServiceID
-	Service       cluster.Service
-	ManifestPath  string
-	ManifestBytes []byte
-	Updates       []update.ContainerUpdate
+type Changes interface {
+	CalculateRelease(update.ReleaseContext, log.Logger) ([]*update.ServiceUpdate, update.Result, error)
+	ReleaseKind() update.ReleaseKind
+	CommitMessage() string
 }
 
-func Release(rc *ReleaseContext, spec update.ReleaseSpec, logger log.Logger) (results update.Result, err error) {
-	started := time.Now()
-	defer func(start time.Time) {
-		releaseDuration.With(
-			fluxmetrics.LabelSuccess, fmt.Sprint(err == nil),
-			fluxmetrics.LabelReleaseType, update.ReleaseSpec(spec).ReleaseType(),
-			fluxmetrics.LabelReleaseKind, string(spec.Kind),
-		).Observe(time.Since(started).Seconds())
-	}(started)
+type Observer interface {
+	Observe(time.Time, error)
+}
 
-	logger = log.NewContext(logger).With("type", "release")
-	// We time each stage of this process, and expose as metrics.
-	var timer *metrics.Timer
+func Release(rc *ReleaseContext, changes Changes, logger log.Logger) (results update.Result, err error) {
+	if o, ok := changes.(Observer); ok {
```

Comment:
This seems destined to cause confusion -- why would changes observe anything? Better to be explicit, either by wrapping this function, or having another argument.